### PR TITLE
[WIP] Add numerics functions

### PR DESCRIPTION
**Pull request by antoinecollas**
WIP: Add numerics functions

This PR follows https://github.com/pymanopt/pymanopt/pull/252

### CODE REVIEW

**Comment by antoinecollas** [comment]
@nkoep I implemented the numpy backend for all the functions you added to the TODO list.

**Comment by antoinecollas** [comment]
@nkoep `examples.closest_unit_norm_column_approximation` is now working with numpy! 

However, we have a problem when calling the method `random_point` of a given manifold. It uses the numpy backend because it is the default one. And it has no clue to figure out which one it should use...

**Comment by antoinecollas** [comment]
The last commit is an attempt to alleviate this problem. By default `random_point` generates a `numpy.ndarray`. However, if the attribute `manifold.backend` is set to something else (pytorch, jax or tensofrflow), then the `numpy.ndarray` is transformed to the appropriate type. The attribute `manifold.backend` is automatically set when defining the problem from the backend used to define the cost function. So it is transparent for the user!

**Comment by antoinecollas** [comment]
@nkoep tell if it is ok for you. Then, I'll fix the tests that now fail and I'll implement `pytorch` backend.

**Comment by antoinecollas** [comment]
Locally, all tests are passing now!

**Comment by nkoep** [comment]
> The last commit is an attempt to alleviate this problem. By default `random_point` generates a `numpy.ndarray`. However, if the attribute `manifold.backend` is set to something else (pytorch, jax or tensofrflow), then the `numpy.ndarray` is transformed to the appropriate type. The attribute `manifold.backend` is automatically set when defining the problem from the backend used to define the cost function. So it is transparent for the user!

Sorry for the delayed response, but I'm not quite following what the issue is. The `random_point` and `random_tangent_vector`  methods just call numpy's `random`  module, don't they? It just means this will also have to be abstracted. What am I missing?

I'll have a look at the PR now.

**Comment by antoinecollas** [comment]
The problem is that 'random_point' does not know which backend it should use. Comparatively, you give a 'point' to 'random_tangent_vector' so it knows which backend it has to use.

**Comment by nkoep** [comment]
> The problem is that 'random_point' does not know which backend it should use. Comparatively, you give a 'point' to 'random_tangent_vector' so it knows which backend it has to use.

Ahh that's true of course. Let's see if we can come with a solution that doesn't make any explicit assumptions about the existence of backends. I don't see how this is supposed to work considering that we override that method in each manifold so the concrete implementations do not see that decorator: https://github.com/pymanopt/pymanopt/pull/253/files#diff-aa7fa40977462154d6b5026e6222de607751d7885ec620e9548ca3df4e490ec1R204.

**Review by nkoep — COMMENTED** [review]
Apart from the hack to solve the type issue of `random_point` implementations this seems to be on the right track. Thanks a lot for all the work, Antoine! I know that it's really tedious and time-consuming to add all these implementations.

File: src/pymanopt/numerics/core.py [hunk 1]
```diff
@@ -12,79 +16,102 @@ def inner(*arguments):
...
+    pass
+
+
+@functools.singledispatch
+@_not_implemented
+def any(_):
     pass
 
 
+arange = np.arange
```
[the remark this answers is not in the document]
Comment (antoinecollas) [reply src/pymanopt/numerics/core.py:43]: How would you do otherwise? The problem is that I want to keep the flexibility of the `np.arange`. If you define it with `functools.singledispatch` then you'll have to provide `start` and `stop`. In practice, often you only use the `stop` argument but sometimes you use both `start` and `stop`. The definition of the API of `np.arange` is quite long in numpy, see https://github.com/numpy/numpy/blob/cdfbdf428d9df9c7119cecae323512a4cd3f57b7/numpy/_core/multiarray.pyi#L712 .

[the remark this answers is not in the document]
Comment (antoinecollas) [reply src/pymanopt/numerics/core.py:43]: And all the functions I directly defined in `core.py` are those that can't be dispatched (like `arange`, `eye`, etc...).

**Comment by antoinecollas** [comment]
Thanks for all the comments Niklas! I'll go through them in the coming week.

File: src/pymanopt/numerics/_backends/numpy.py [hunk 2]
```diff
@@ -1,30 +1,290 @@
 import numpy as np
+import scipy
 
-import pymanopt.numerics.core as nx
+import pymanopt.numerics as nx
+
+generic_np_type = int | float | complex | np.generic | tuple | list | np.ndarray
```
[the remark this answers is not in the document]
Comment (antoinecollas) [reply src/pymanopt/numerics/_backends/numpy.py]: I agree but using `typing.Union` in `singledispatch` is new in 3.11. So it does not solve the problem... 
https://docs.python.org/3/whatsnew/3.11.html#functools

And it raises an error when you use `np.typing.ArrayLike` in `singledispatch`.

File: src/pymanopt/numerics/_backends/numpy.py [hunk 3]
```diff
@@ -1,30 +1,290 @@
...
     return np.allclose(array_a, array_b)
 
 
+@nx.any.register
+def _(array: generic_np_type) -> bool:
+    return np.any(array)
+
+
+@nx.arccos.register
+def _(array: generic_np_type) -> np.float64 | np.ndarray:
```
[the remark this answers is not in the document]
Comment (antoinecollas) [reply src/pymanopt/numerics/_backends/numpy.py]: Or maybe remove the output type hinting?

**Comment by antoinecollas** [comment]
@nkoep I fixed the issue with the `random_point` method by creating the metaclass BackendManifold that wraps `random_point`.

**Comment by antoinecollas** [comment]
I'll try to get an example like `dominant_eigenvector.py` working end to end with `torch`.

**Comment by antoinecollas** [comment]
@nkoep, I changed quite a lot of things, but we are getting close to having the first version of the multi-backend support for manifolds operations!

For now, I focused on having this working for numpy and torch. This means you can still use the 4 backends (numpy, torch, jax and tensorflow) to define your loss function but for jax and tensorflow, operations on the manifold are still performed with numpy. For Pytorch, all operations are performed with Pytorch and there are no more conversions to numpy arrays. The list of available backends for numericals is available at `nx.NUMERICS_SUPPORTED_BACKENDS`.

To get an overview of this PR, you can begin with `examples/dominant_eigenvector.py`. It is nearly identical as before expect the point returned by the optimiser is now a `torch.tensor` when using PyTorch backend.

The manifolds are backend agnostic, i.e. they instantiated without specifying the backend of input data. The only case when the manifold must know the backend is when sampling a point with the `random_point` method. In this case, the user need to set `manifold.backend` if he wants something else than a numpy array. Indeed, by default, a given manifold samples points with numpy.

When running `optimizer.run(problem)`, the backend is inferred from the decorator of the loss function `@pymanopt.function.XXX(manifold)`.

All manifolds are still tested with numpy. `Sphere` is tested with all `nx.NUMERICS_SUPPORTED_BACKENDS`, i.e `numpy` and `PyTorch` thanks to `TestBackendManifoldFactory`. The only change in `Sphere` is the `random_tangent_vector` where an `nx.array_as` is added to change the backend of the generated `vector`. `random_point` is not changed since it is decorated by `Manifold`, which automatically changes the backend of the generated data.

When a manifold stores data like in `_SphereSubspaceIntersectionManifold`,  then data are now stored in `self.parameters`. The latter automatically changes the backend of what is stored inside when `manifold.backend` is changed.

Some modules had to be adapted such as `tools/multi.py` and `tools/testing.py`. `tools/multi.py` is fully tested with all `nx.NUMERICS_SUPPORTED_BACKENDS`.

Overall, most of the remaining work concerns the tests of all manifolds except `Sphere`.

@nkoep, tell me what you think about all of this!
